rascal-lsp/src/main/rascal/lang/rascal/tests/rename/Fields.rsc

```diff
@@ -0,0 +1,173 @@
+    '   f = a(8);
+    '   g = a.foo;
+    '}
+    ", {0})
+}, <"Main", "foo", 0>);
+
+test bool extendedDataField() = testRenameOccurrences({
+    byText("Scratch1", "
+        'data Foo = f(int foo);
+        ", {0}),
+    byText("Scratch2", "
+        'extend Scratch1;
+        'data Foo = g(int foo);
+        ", {0})
+}, <"Scratch2", "foo", 0>);
+
+test bool relFieldAtDef() = {0, 1} == testRenameOccurrences("
+    'rel[str foo, str baz] r = {};
+    'f = r.foo;
+");
+
+test bool relFieldAtUse() = {0, 1} == testRenameOccurrences("
+    'rel[str foo, str baz] r = {};
+    'f = r.foo;
+", cursorAtOldNameOccurrence = 1);
+
+test bool lrelFieldAtDef() = {0, 1} == testRenameOccurrences("
+    'lrel[str foo, str baz] r = [];
+    'f = r.foo;
+");
+
+test bool lrelFieldAtUse() = {0, 1} == testRenameOccurrences("
+    'lrel[str foo, str baz] r = [];
+    'foos = r.foo;
+", cursorAtOldNameOccurrence = 1);
+
+test bool relSubscript() = {0, 1} == testRenameOccurrences("
+    'rel[str foo, str baz] r = {};
+    'x = r\<foo\>;
+");
+
+test bool relSubscriptWithVar() = {0, 2} == testRenameOccurrences("
+    'rel[str foo, str baz] r = {};
+    'str foo = \"foo\";
+    'x = r\<foo\>;
+");
+
+test bool tupleFieldSubscriptUpdate() = {0, 1, 2} == testRenameOccurrences("
+    'tuple[str foo, int baz] t = \<\"one\", 1\>;
+    'u = t[foo = \"two\"];
+    'v = u.foo;
+");
+
+test bool tupleFieldAccessUpdate() = {0, 1} == testRenameOccurrences("
+    'tuple[str foo, int baz] t = \<\"one\", 1\>;
+    't.foo = \"two\";
+");
+
+test bool similarCollectionTypes() = {0, 1, 2, 3, 4} == testRenameOccurrences("
+    'rel[str foo, int baz] r = {};
+    'lrel[str foo, int baz] lr = [];
+    'set[tuple[str foo, int baz]] st = {};
+    'list[tuple[str foo, int baz]] lt = [];
+    'tuple[str foo, int baz] t = \<\"\", 0\>;
+");
+
+test bool differentRelWithSameFieldAtDef() = {0, 1} == testRenameOccurrences("
+    'rel[str foo, int baz] r1 = {};
+    'foos1 = r1.foo;
+    'rel[int n, str foo] r2 = {};
+    'foos2 = r2.foo;
+");
+
+test bool differentRelWithSameFieldAtUse() = {0, 1} == testRenameOccurrences("
+    'rel[str foo, int baz] r1 = {};
+    'foos1 = r1.foo;
+    'rel[int n, str foo] r2 = {};
+    'foos2 = r2.foo;
+", cursorAtOldNameOccurrence = 1);
+
+test bool tupleFieldAtDef() = {0, 1} == testRenameOccurrences("
+    'tuple[int foo] t = \<8\>;
+    'y = t.foo;
+");
+
+test bool tupleFieldAtUse() = {0, 1} == testRenameOccurrences("
+    'tuple[int foo] t = \<8\>;
+    'y = t.foo;
+", cursorAtOldNameOccurrence = 1);
+
+// We would prefer an illegalRename exception here
+@expected{unsupportedRename}
+test bool builtinFieldSimpleType() = testRename("
+    'loc l = |unknown:///|;
+    'f = l.top;
+", oldName = "top", newName = "summit");
+// We would prefer an illegalRename exception here
+
+@expected{unsupportedRename}
+test bool builtinFieldCollectionType() = testRename("
+    'loc l = |unknown:///|;
+    'f = l.ls;
+", oldName = "ls", newName = "contents");
```

Comment:
Suggestion: consider the following fragment.
```
data D = d0(int foo) | d1(str bar);
```
The renaming should be forbidden, as field names in an ADT should not be given to fields of different types.

Reply:
Will address ADTs in a future PR.